examples/other_files/ML_AI_Plugin/mea_column_model_training_jenn.py

```diff
@@ -0,0 +1,151 @@
+            epochs=1,  # number of passes through data
+            batch_size=None,  # used to divide data into training batches (use for large data sets)
+            max_iter=200,  # number of optimizer iterations per mini-batch
+            shuffle=True,
+            random_state=None,
+            is_backtracking=False,
+            is_verbose=False,
+        )
+
+        y_pred = np.transpose(model.predict(np.transpose(x_train)))
+        SSE = sum((y_pred - y_train) ** 2)
+
+        # y0 is 2.1 orders of magnitude larger than y1, so adjust the SSE check
+        # CO2 capture rate and SRD should both be positive for all predictions
+        # SSE =  [75589.3371621  13214.64474031] from running 1000
+        # SSE =  [39801.73811642  436.51381078] from running 20000
+        if (SSE[0] / 21 + SSE[1]) < (best_SSE[0] / 21 + best_SSE[1]) and np.all(
+            y_pred
+        ) > 0:
+            best_SSE = SSE
+            best_model = model
+            best_y_pred = y_pred
+
+        print(np.round(idx / runs * 100, 3), " % complete")
+
+    best_model.custom = SimpleNamespace(
+        input_labels=xlabels,
+        output_labels=zlabels,
+        input_bounds=xdata_bounds,
+        output_bounds=ydata_bounds,
+        normalized=False,  # JENN models are normalized during training, this should always be False
+    )
+
+    return best_model, best_y_pred, best_SSE
+
+
+# Main code
+
+# import data
+data = pd.read_csv(r"MEA_carbon_capture_dataset_mimo.csv")
+grad0_data = pd.read_csv(r"gradients_output0.csv", index_col=0)  # ignore 1st col
+grad1_data = pd.read_csv(r"gradients_output1.csv", index_col=0)  # ignore 1st col
+
+xdata = data.iloc[:, :6]  # there are 6 input variables/columns
+ydata = data.iloc[:, 6:]  # the rest are output variables/columns
+xlabels = xdata.columns.tolist()  # set labels as a list (default) from pandas
+zlabels = ydata.columns.tolist()  # is a set of IndexedDataSeries objects
+xdata_bounds = {i: (xdata[i].min(), xdata[i].max()) for i in xdata}  # x bounds
+ydata_bounds = {j: (ydata[j].min(), ydata[j].max()) for j in ydata}  # y bounds
+
+xmax, xmin = xdata.max(axis=0), xdata.min(axis=0)
+ymax, ymin = ydata.max(axis=0), ydata.min(axis=0)
+xdata, ydata = np.array(xdata), np.array(ydata)  # (n_m, n_x) and (n_m, n_y)
+gdata = np.stack([np.array(grad0_data), np.array(grad1_data)])  # (2, n_m, n_x)
+
+model_data = np.concatenate(
+    (xdata, ydata), axis=1
+)  # JENN requires a Numpy array as input
+
+# define x and y data, not used but will add to variable dictionary
+xdata = model_data[:, :-2]
+ydata = model_data[:, -2:]
+
+# create model
+model, y_pred, SSE = create_model(x_train=xdata, y_train=ydata, grad_train=gdata)
+
+with open("mea_column_model_jenn.pkl", "wb") as file:
+    pickle.dump(model, file)
+
+# load model as pickle format
+with open("mea_column_model_jenn.pkl", "rb") as file:
+    loaded_model = pickle.load(file)
+
+
+print(y_pred)
+print("SSE = ", SSE)
```

Comment:
Would it be possible to wrap this code in a `main()` function, so that it's not run on import?

```py

def main():
    # move code here


if __name__ == "__main__":
    main()
```

Please disregard if the current structure is required (e.g. the code needs to be at the module scope for the dynamic module import to work correctly).

Reply:
That should be possible for the code that calls the training method, as all of the training examples are only used to generate the model files. The training methods that create the models may need to be on the module level - Keras for sure needs this to register and load the model class object.

Some of the Keras example files are vehicles for the class objects, e.g. `mea_column_model_customnormform.py` must be copied to the working directory during testing to load the registered model class in the plugin code, and those non-training files don't have any "main" code to execute.

I can update that here, or in another PR.